iree/modules/strings/strings_module.cc

```diff
@@ -67,23 +50,158 @@ class StringsModuleState final {
   Status Initialize() { return OkStatus(); }
 
   // strings.print(%str)
-  Status Print(vm::ref<vmstring_t> str) {
+  Status Print(vm::ref<string_t> str) {
     fwrite(str->value.data, 1, str->value.size, stdout);
     fputc('\n', stdout);
     fflush(stdout);
     return OkStatus();
   }
 
   // strings.i32_to_string(%value) -> %str
-  StatusOr<vm::ref<vmstring_t>> I32ToString(int32_t value) {
-    vm::ref<vmstring_t> new_string;
+  StatusOr<vm::ref<string_t>> I32ToString(int32_t value) {
+    vm::ref<string_t> new_string;
     std::string str = std::to_string(value);
     RETURN_IF_ERROR(
-        FromApiStatus(vmstring_create(iree_make_cstring_view(str.c_str()),
-                                      allocator_, &new_string),
+        FromApiStatus(string_create(iree_make_cstring_view(str.c_str()),
+                                    allocator_, &new_string),
                       IREE_LOC));
+    return new_string;
+  }
+
+  const iree_string_view_t* StringTensorToStringHelper(
+      const iree_string_view_t* strs, const int32_t* shape, int32_t rank,
+      std::string* output) {
+    // Handle a scalar tensor value.
+    if (rank == 0) {
+      const auto& str = strs[0];
+      output->append(str.data, str.size);
+      return strs + 1;
+    }
+
+    // The row for the final tensor dimension.
+    if (rank == 1) {
+      output->append("[", 1);
+      for (int32_t i = 0, s = shape[0]; i < s; i++) {
+        const auto& str = strs[i];
+        output->append(str.data, str.size);
+        if (i != s - 1) {
+          output->append(", ", 2);
+        }
+      }
+
+      output->append("]", 1);
+      return strs + shape[0];
+    }
+
+    // Recurse to the lower dimension with the approrpiate brackets.
+    output->append("[", 1);
+    for (int32_t i = 0, s = shape[0]; i < s; i++) {
+      strs = StringTensorToStringHelper(strs, shape + 1, rank - 1, output);
+      if (i != s - 1) {
+        output->append(",\n", 2);
+      }
+    }
+    output->append("]", 1);
+    return strs;
+  }
 
-    return std::move(new_string);
+  // strings.print_tensor(%str_tensor)
+  StatusOr<vm::ref<string_t>> StringTensorToString(
+      vm::ref<string_tensor_t> str_tensor) {
+    size_t string_length = 0;
+    for (int i = str_tensor->rank - 1; i >= 0; i--) {
```

Comment:
This seems like highly error-prone code that is coupled to StringTensorToStringHelper, which is only used for the reserve call (pure optimization). Let's remove it for now. We can re-add it as an optimization later if needed, and maybe do something simpler.

Reply:
I tweaked it to a simplified version. It may still result in reallocations but it should handle the base cases.